Tenant Token (#89)
meilisearch/specifications: text/0089-scoped-api-keys.md

- Title: Scoped API Keys
- Start Date: 2021-10-15
- Specification PR: [#89](https://github.com/meilisearch/specifications/pull/89)
- Discovery Issue: [#51](https://github.com/meilisearch/product/issues/51)

# Scoped API Keys

## 1. Functional Specification

### I. Summary

> The SDKs can generate `Scoped API Keys` inheriting from a MeiliSearch's `API key` generated to force filters during a search for an end-user. A `Scoped API Key` is generated on the user server-side code to be used by an end-user. It allows users to have multi-tenant indexes and thus restricts access to documents depending on the end-user making the search request.

Reviewer: >  generated to force filters during a search for an end-user

This is confusing. What do you mean?

> A `Scoped API Key` is generated on the user server-side code

Do you mean the request to generate is made at the server side? Or is MeiliSearch instance the server side?

> depending on the end-user making the search request

I wouln't use `making the search request` as you may create a scoped API key for users managing indexes and not specifically for search requests.

Author: > generated to force filters during a search for an end-user

Q: This is confusing. What do you mean?
A: It means that a `Scoped API Key` is forged with a specific `filter` parameter that cannot be deleted. They are inserted in the `Scoped API Key` payload. When the client-code emits a search request, the search engine will first validate the authorization of the `Scoped API Key`, extract the Scoped API Key `filter` value and add the search request `filter` parameter behind an `AND` operator.

> A Scoped API Key is generated on the user server-side code

Q: Do you mean the request to generate is made at the server side? Or is MeiliSearch instance the server side?
A: It is generated by calling the SDK method on the server-side user code. MeiliSearch receives nothing and stores nothing when generating a Scoped API Keys when a user generates one for an end-user.

> depending on the end-user making the search request

Q: I wouldn't use making the search request as you may create a scoped API key for users managing indexes and not specifically for search requests.
A: For now, a `Scoped API Key` can only be used for a search request. Using a Scoped API Key to reach an endpoint other than /search will raise an `invalid_api_key` error. We can improve this in the future! 🤓

### II. Motivation

> `Scoped API Keys` are introduced to solve multi-tenant use-case. By managing this, we reduce one of the last major deal-breakers that makes users not choose MeiliSearch as a solution despite all our advantages.

Reviewer: ```suggestion
`Scoped API Keys` are introduced to solve multi-tenant use-cases. By providing the management of API Keys, we remove one of the last major deal-breakers that makes users not choose MeiliSearch as a solution despite all our advantages.
```

Users regularly request Multi-Tenant indexes. Users today need to set up workarounds to meet this need. Some of them implement reverse-proxy or managed authentication systems like Hasura or Kong to filter what can and cannot be read. Others decide to use server code as a facade to implement the access restriction logic before requesting MeiliSearch. It is difficult to maintain, less efficient, and requires necessary skills that not everyone has.

### III. Glossary

| Term               | Definition |
|--------------------|------------|

> | Main Key         | This is the main key that allows the creation of other API keys. The main key is defined by the user when launching MeiliSearch. |

Reviewer: Will master key be renamed to main key?

Author: Absolutely! This is specified in https://github.com/meilisearch/specifications/pull/85.

> | API Key            | API keys are stored and managed from the endpoint `/keys` by the main key holder. These are the keys used by the technical teams to interact with MeiliSearch at the level of the client code. |

Reviewer: >  to interact with MeiliSearch at the level of the client code.

What do you mean with this?

Author: Client code here means anything that can communicate and interact with MeiliSearch. A curl script, an SDK etc

| Scoped API Key     | These keys are not stored and managed by a MeiliSearch instance. They are generated for each end-user by the backend code from a MeiliSearch API Key. They are used by the end-users to only search the documents belonging to them. |

> | Multi-Tenancy      | By multi-tenancy, we mean that an end-user only accesses data belonging to him within an index shared with other end-users. |

Reviewer: ```suggestion
| Multi-Tenancy      |  An instance that only accesses a limited set of data within an index shared with other end-users. |
```

### IV. Personas

| Persona | Role |
|---------|------|
| Mark    | Mark is a developer for a SaaS company. He will implement the code to communicate with MeiliSearch to solve technical/product needs. |
| UserX   | UserX represents any end-user searching from the frontend interfaces provided by Anna and Mark's company. |

### V. `Scoped API Key` Explanations

#### Summary Key Points

- `Scoped API keys` are generated from a MeiliSearch parent `API key` on the client's server code to resolve multi-tenancy use-case by restricting access to data within an index according to the criteria chosen by the team managing a MeiliSearch instance.
- These `Scoped API keys` can't be less restrictive than a parent `API key` and can only be used for the search action with a predefined forced filter field.
- These `Scoped API keys` are not stored and thus retrievable on MeiliSearch. This is why we highly advise setting an expiration time on that type of API key for security reasons.

#### Solving Multi-Tenancy with `Scoped API Keys`

![](https://i.imgur.com/J4jVe1n.png)

> Let's say that `Mark` is a developer for a SaaS platform. He would like to ensure that an end-user can only access his documents at search time. **His database contains many users and he hopes to have many more in the future.**

Reviewer: ```suggestion
Let's say that `Mark` is a developer for a SaaS platform. He would like to ensure that an end-user can only access their documents at search time. **His database contains many users and he hopes to have many more in the future.**
```

> When a user registers, the backend-side client code generates a `Scoped API Key` specifically for that end-user so he can only access his documents.

Reviewer: ```suggestion
When a user registers, the backend-side client code generates a `Scoped API Key` specifically for that end-user so they can only access their documents.
```

> A `filter` parameter is set to restrict the search for documents having a `user_id` attribute. This `filter` parameter is contained in the `Scoped API key` payload and cannot be modified during the search by the end-user making the request. `filter` can be made of any valid filters. e.g. `user_id = 10 and category = Romantic`

Reviewer: ```suggestion
The `filter` parameter is set to restrict the search for documents having an `user_id` attribute. This `filter` parameter is contained in the `Scoped API key` payload and cannot be modified during the search by the end-user making the request. `filter` can be made of any valid filters. e.g. `user_id = 10 and category = Romantic`
```

> This `Scoped API Key` is generated from a parent `API key` used to cipher the `Scoped API Key`. On the MeiliSearch side, this permits checking permissions and authorizing the `UserX` search request using this `Scoped API Key`.

Reviewer: ```suggestion
This `Scoped API Key` is generated from a parent `API key` used to cipher the `Scoped API Key`. On the MeiliSearch side, the payload of the `scoped API key` defines what data the user is allowed to retrieve, in order words, MeiliSearch grants authorization based on the permission defined in the Scoped API key payload.
```

---

#### Generating a `Scoped API Key`

```javascript
const scopedApiKeyRestrictions = {
     "indexesPolicy": {
         "products": {
             "filter": "user_id = 1"
         },
         "reviews": {

>              "filter": "user_id = 1 AND published = true"

Reviewer: This example is a bit confusing as `products` do not have an associated `user_id` normally. Since we are not a relational database I'm not sure what use cases it represents

Author: Imagine that you have 2 indexes: `products` and `reviews` and you only want end-users to retrieve documents containing `user_id = :userId`. You also want to ensure that end-users will only access published reviews when searching documents from the `reviews` index.

         }
     },
    "expiresIn": 3600
}

export const generateScopedApiKey = () => {
  return (parentApiKey: string, restrictions: scopedApiKeyRestrictions): string => {
    //hash the parentApiKey and keep a prefix of 4 chars of the hashed parentApiKey
    const prefixKey = crypto
        .createHmac('sha256', parentApiKey)
        .digest('hex')
        .substr(0, 4);

    //serialize restrictions (indexesPolicies object and expiresIn)
    const queryParameters = serializeQueryParameters(restrictions);

    //create the secured part (parentApiKey + queryParameters)
    const securedKey = crypto
      .createHmac('sha256', parentApiKey)
      .update(queryParameters)
      .digest('hex');

    //return the generated `Scoped Api Key`
    return Buffer.from(prefixKey + securedKey + queryParameters).toString('base64');
  };

> };

Reviewer: Is this expected from the user?

Author: This is the pseudo-code for an SDK, `meilisearch-js` in this case.

When the user will want to create a `Scoped API Key` for an end-user, his server code will have to call the `generateScopedApiKey` method and give it to the frontend code to authorize the end user's search requests.

```

##### scopedApiKeyRestrictions

The format allows defining specific enforced search filters for accessible indexes (these indexes must be defined in the parent `Api Key` used to generate the `Scoped API key` and have the search action).

> If the user does not want to define specific filters for each index accessible to the end-user, he can use the `*` index wildcard rule.

Reviewer: Some indexes may not be accessible?

Author: The accessible indexes are those defined by the parent `API Key` used (which must have the `search` action permission) to generate the `Scoped API Key`. This format here allows setting a filter on particular fields if there is no common field between documents coming from different accessible indexes or to force some particular `filter` rule for specific indexes.
